numpy/random/tests/test_random.py

```diff
@@ -510,6 +512,31 @@ def test_shuffle_masked(self):
             assert_equal(
                 sorted(b.data[~b.mask]), sorted(b_orig.data[~b_orig.mask]))
 
+    @pytest.mark.parametrize("random",
+        [np.random, np.random.RandomState(), np.random.default_rng()])
+    @pytest.mark.parametrize("use_array_like", [True, False])
+    def test_shuffle_no_object_unpacking(self, random, use_array_like):
+        class MyArr(np.ndarray):
+            pass
+
+        items = [None, np.array([3]), np.float64(3), np.array(10), np.float64(7)]
+        arr = np.array(items, dtype=object)
+        item_ids = {id(i) for i in items}
+        if use_array_like:
+            arr = arr.view(MyArr)
+
+        # The array was created fine, and did not modify any objects:
+        assert all(id(i) in item_ids for i in arr)
+
+        if use_array_like and not isinstance(random, np.random.Generator):
+            # The old API gives incorrect results, but warns about it.
```

Comment:
```suggestion
            # The old API gives incorrect results (duplicating and mutating elements), but warns about it.
```

Reply:
I wonder whether it's worth including the failure in a comment - I got:
```
>>> orig_arr
array([None, np.array([3]), np.float64(3), np.array(10), np.float64(7)], dtype=object)
>>> arr
MyArr([7.0, MyArr(None, dtype=object), 3.0, array(10), MyArr(None, dtype=object)], dtype=object)
```
which is _very_ different from the original array.

Reply:
commited your suggestions (with added line breaks).  Well, the code in the test is a (simple) example, or did you mean in the warning?